### PR TITLE
docs(ADR): #113 Modular transaction hashing

### LINKED ISSUE #342: ADR for modular transaction hashing
CometBFT should support pluggable cryptographic functionality in order to allow users to address a wider range of use cases than possible with the default cryptography we provide. At present the only way they can do this is to fork the project.

Specific use cases here are to be clarified so the team understands what kinds of interface(s) to implement. Another concern potentially addressed by this is one of mitigating security-related risk through isolation of cryptography-related dependencies (probably our highest-impact dependencies) in order to make them easier to replace in case of vulnerabilities.

The first thing we need here is a design in the form of an ADR, and it must be clear from the ADR as to how the aforementioned use cases could be implemented using the design.

### PR DESCRIPTION
**Pull request by melekes**
[RENDERED](https://github.com/cometbft/cometbft/blob/anton/342-txi/docs/references/architecture/adr-113-modular-transaction-hashing.md)

Closes #342

### CODE REVIEW

**Comment by adizere** [comment]
@itsdevbear @fedekunze @JayT106 super simple ADR, please have a look.

**Review by adizere — APPROVED** [review]
Concept ACK, thanks Anton!

I pinged a couple of users to gather some feedback on this, let's wait on their input before merging.

**Comment by cason** [comment]
Hey, before reviewing this, do you mind using 113 for this ADR? We are working in the PBTS feature branch on an ADR 112.

File: docs/references/architecture/adr-112-txi.md
Comment (cason) [inline docs/references/architecture/adr-112-txi.md:13]:
```suggestion
Transaction hashing in CometBFT is currently implemented using `crypto/tmhash`
```

File: docs/references/architecture/adr-112-txi.md
Comment (cason) [inline docs/references/architecture/adr-112-txi.md:14]:
```suggestion
package, which itself relies on `[sha256](https://pkg.go.dev/crypto/sha256)` to calculate a transaction's hash.
```

File: docs/references/architecture/adr-112-txi.md
Comment (cason) [inline docs/references/architecture/adr-112-txi.md:28]:
```suggestion
In order to be flexible, CometBFT needs to allow changing the transaction hashing algorithm
```

File: docs/references/architecture/adr-112-txi.md
Comment (cason) [inline docs/references/architecture/adr-112-txi.md:16]: Can you elaborate more this list?

The mempool implementation uses transaction hashes. Both to produce log messages and as a key to maps of transactions (`types.Tx.Key()` is a hash as an array with fixed size).

File: docs/references/architecture/adr-112-txi.md
Comment (cason) [inline docs/references/architecture/adr-112-txi.md:21]: Not true for commit, header and partset header. 

They use the `merkle` package which, it is true, uses `sha256`.

File: docs/references/architecture/adr-112-txi.md
Comment (cason) [inline docs/references/architecture/adr-112-txi.md:26]: Do we have an issue or a reference for that?

File: docs/references/architecture/adr-112-txi.md
Comment (cason) [inline docs/references/architecture/adr-112-txi.md:37]:
```suggestion
Give app developers a way to provide their own transaction hash function.
```

File: docs/references/architecture/adr-112-txi.md
Comment (cason) [inline docs/references/architecture/adr-112-txi.md:64]: Can you be more specific regarding that?

In the case of `types/Tx`, calculating the hash is hard-coded. How can we parameterize this?

Comment (cason) [reply docs/references/architecture/adr-112-txi.md:64]: In the case we want this change to apply to the RPC endpoints and, possibly, indexer is one thing. In this case we can use any customized hash function.

But if we want to adopt customized hash functions for our types, this is a completely different discussion.

File: docs/references/architecture/adr-112-txi.md
Comment (cason) [inline docs/references/architecture/adr-112-txi.md:76]: Not really, it can be an `Option`: https://github.com/cometbft/cometbft/blob/main/node/node.go#L96

File: docs/references/architecture/adr-112-txi.md
Comment (cason) [inline docs/references/architecture/adr-112-txi.md:78]: Not sure if performance should be considered here at some of the consequences? Probably neutral, since it really depends on the hash method adopted.

**Review by cason — COMMENTED** [review]
Nice write up.

I agree with the approach, but I think we should really restrict the scope of the proposed changes. I know, for instance, that the RPC endpoints use some weird way to "print" (i.e., representing as strings) transactions and their IDs/hashes. But this is a problem of `String()` method. That I am afraid that is not solved using this approach.

**Comment by cason** [comment]
Recall also to add an entry to the `README.md` file in the same directory.

File: docs/references/architecture/adr-112-txi.md
Comment (melekes) [reply docs/references/architecture/adr-112-txi.md]: > Not true for commit, header and partset header.
> They use the merkle package which, it is true, uses sha256.

https://github.com/cometbft/cometbft/blob/d02b25e0fd278bc7e70142190ae5b3b8e722a8df/crypto/merkle/hash.go

are you sure?

**Review by andynog — COMMENTED** [review]
The overall approach seems OK, but my concern is related to the scope and implications of this change. 

For example, in `types/tx.go` the code below might affect the Merkle root hash
```
func (txs Txs) Hash() []byte {
	hl := txs.hashList()
	return merkle.HashFromByteSlices(hl)
}
```

because in `txs.hashList()` the `Hash()` is invoked
```
func (txs Txs) hashList() [][]byte {
	hl := make([][]byte, len(txs))
	for i := 0; i < len(txs); i++ {
		hl[i] = txs[i].Hash()
	}
	return hl
}
```
which in `merkle.HashFromByteSlices(hl)` will make the Merkle root hash different
```
func HashFromByteSlices(items [][]byte) []byte {
	return hashFromByteSlices(sha256.New(), items)
}
```
maybe this will be OK, but will this have implications later on ? Like during an upgrade, could the hash algorithm be changed ? what are the implications on verifying and validating if the Merkle is not valid.

Also, is this OK with IBC and proofs ? Like what if one chain decides to switch the hash algorithm? what are the validity of proof guarantees ? 

Anyway, the scope and implications of this change seems to be a lot broader and it is a good idea to think this through. Maybe my concerns are not valid but just bringing this up 😉

File: docs/references/architecture/adr-112-txi.md
Comment (cason) [reply docs/references/architecture/adr-112-txi.md]: True, in the verification side, sorry.

**Comment by cason** [comment]
> The overall approach seems OK, but my concern is related to the scope and implications of this change.

Fully agree with that, much better explained by @andynog. Hashing transactions is something useful for the mempool reactor and for RPC endpoints.

Other uses of this library includes block, merkle tree, evidence production and verification. For this case, being compatible is essential. A suggestion here, for instance, is to adopt `ed25519` directly, so it is clear to everyone that this is the standard for blockchain hashes.

File: docs/references/architecture/adr-113-txi.md
Comment (lasarojc) [inline docs/references/architecture/adr-113-txi.md:27]: Is there any need to also RLP encoding or really just the hasn?

**Comment by robert-zaremba** [comment]
NOTE: We should be able to customize the Merkle Tree hashing as well. But this has to be done in the IAVL level.

File: docs/references/architecture/adr-113-txi.md
Comment (cason) [inline docs/references/architecture/adr-113-txi.md:14]:
```suggestion
package, which itself relies on [`sha256`](https://pkg.go.dev/crypto/sha256) to calculate a transaction's hash.
```

File: docs/references/architecture/adr-113-txi.md
Comment (cason) [inline docs/references/architecture/adr-113-txi.md:41]: Should we be specific and add Transaction or Tx to the name and description of the methods?

File: docs/references/architecture/adr-113-txi.md
Comment (cason) [inline docs/references/architecture/adr-113-txi.md:45]: Doesn't this interface needs a `Sum()` method?

File: docs/references/architecture/adr-113-txi.md
Comment (cason) [inline docs/references/architecture/adr-113-txi.md:69]: This is `Sum`, I think.

Missing `Size() int`.

File: docs/references/architecture/adr-113-txi.md
Comment (cason) [inline docs/references/architecture/adr-113-txi.md:74]: Can we get a list of where in the code we need to use this hash function? And which interfaces we would need to change?

File: docs/references/architecture/adr-113-txi.md
Comment (cason) [inline docs/references/architecture/adr-113-txi.md:85]: Added complexity? Not really a problem, but a negative side.

Comment (adizere) [reply docs/references/architecture/adr-113-txi.md:85]: agree

9ee941ab2

File: docs/references/architecture/adr-113-txi.md
Comment (adizere) [reply docs/references/architecture/adr-113-txi.md:27]: Judging from the [Evmos fork and initial investigation in this comment](https://github.com/cometbft/cometbft/issues/342#issuecomment-1888821207), encoding was not a concern.

**Comment by adizere** [comment]
It would be good if we can demonstrate that the approach in this ADR can achieve the same effects as in https://github.com/evmos/cometbft/pull/1.

File: docs/references/architecture/adr-113-txi.md
Comment (fedekunze) [inline docs/references/architecture/adr-113-txi.md:47]: what is a `hash.Hash`? It's not defined anywhere. 

Is it an interface? We need to use geth's `common.Hash` which has 32 bytes of size and formats the hash using `0x....` instead of using uppercase of the hex value without the `0x` prefix

File: docs/references/architecture/adr-113-txi.md
Comment (fedekunze) [inline docs/references/architecture/adr-113-txi.md:53]: what's `Option`? I assume it's for the node but needs to be defined

Comment (fedekunze) [reply docs/references/architecture/adr-113-txi.md:53]: it's also not clear where you connect this custom hashing function to CometBFT

File: docs/references/architecture/adr-113-txi.md
Comment (fedekunze) [inline docs/references/architecture/adr-113-txi.md:45]: we should also have a custom `String` for formatting the hash result

File: docs/references/architecture/adr-113-txi.md
Comment (fedekunze) [inline docs/references/architecture/adr-113-txi.md:66]: this needs to be more explicit. Consider renaming the function to `Hash() hash.Hash`

File: docs/references/architecture/adr-113-txi.md
Comment (fedekunze) [reply docs/references/architecture/adr-113-txi.md:74]: agree, this is missing

File: docs/references/architecture/adr-113-txi.md
Comment (fedekunze) [inline docs/references/architecture/adr-113-txi.md:65]: let's add an example with the ethereum tx hash:

```go
// EVMTxRLPHash uses RLP from an Ethereum tx to implement transaction hashing.
type EVMTxRLPHash struct {
   tx *types.Transaction
}

func (h EVMTxRLPHash) Hash() hash.Hash {
   return h.tx.Hash()
}

func (h EVMTxRLPHash) String() string {
   return h.tx.Hash().String()
}

func (EVMTxRLPHash) Size() int {
  return common.HashLength
}
```

File: docs/references/architecture/adr-113-txi.md
Comment (cason) [reply docs/references/architecture/adr-113-txi.md:47]: This is a go standard interface: https://pkg.go.dev/hash#Hash

Comment (fedekunze) [reply docs/references/architecture/adr-113-txi.md:47]: I see. The `Sum(b []byte) []byte` function from `hash.Hash` is not relevant in the case of Ethereum txs because the hash is computed from the transaction fields and not the transaction bytes. Unless the bytes are unmarshalled to the `Transaction` struct and from there call `tx.Hash()`

File: docs/references/architecture/adr-113-txi.md
Comment (fedekunze) [inline docs/references/architecture/adr-113-txi.md:13]: Is this ADR meant only for transactions or also for the `Block` hash?

File: docs/references/architecture/adr-113-txi.md
Comment (fedekunze) [inline docs/references/architecture/adr-113-txi.md:45]: Not, should be renamed to `TxHashFn` for clarity

File: docs/references/architecture/adr-113-txi.md
Comment (cason) [reply docs/references/architecture/adr-113-txi.md:13]: At the moment, only for transaction's hash. I would make it very precise in the description.

**Comment by adizere** [comment]
blocked on #2610

**Comment by andynog** [comment]
> blocked on #2610

#2610 got merged, so I'm assuming this is good to merge @adizere and @melekes ? I fixed a small conflict so should pass all the tests now

**Comment by adizere** [comment]
> > blocked on #2610
> 
> #2610 got merged, so I'm assuming this is good to merge @adizere and @melekes ? I fixed a small conflict so should pass all the tests now

I think we have two options:
1. merge this, open a new PR with a new ADR 114 (?) which fixes [Fede's comments](https://github.com/cometbft/cometbft/pull/2241#pullrequestreview-1920405754), mark ADR 113 as 'not adopted' and the new ADR 114 as 'proposed' or something similar
2. fix the present ADR to address Fede's comments

probably option 2 is faster and better, so I would suggest doing that

**Review by adizere — DISMISSED** [review]
Fix Fede's & Daniels comments

File: docs/references/architecture/adr-112-txi.md
Comment (melekes) [reply docs/references/architecture/adr-112-txi.md]: > But if we want to adopt customized hash functions for our types, this is a completely different discussion.

I'd like for us to have both discussions.

File: docs/references/architecture/adr-113-txi.md
Comment (melekes) [reply docs/references/architecture/adr-113-txi.md]: If we decide to restrict the scope to just txs, yes.

File: docs/references/architecture/adr-113-txi.md
Comment (melekes) [reply docs/references/architecture/adr-113-txi.md]: Fixed by using a global variable.

File: docs/references/architecture/adr-113-modular-transaction-hashing.md
Comment (sergio-mena) [inline docs/references/architecture/adr-113-modular-transaction-hashing.md:55]:
```suggestion
	"hash"
```

File: docs/references/architecture/adr-113-txi.md
Comment (melekes) [reply docs/references/architecture/adr-113-txi.md]: Fixed in new version.

> what's Option?

https://pkg.go.dev/github.com/cometbft/cometbft@v0.38.7/node#Option

File: docs/references/architecture/adr-113-txi.md
Comment (melekes) [reply docs/references/architecture/adr-113-txi.md]: Done

File: docs/references/architecture/adr-113-txi.md
Comment (melekes) [reply docs/references/architecture/adr-113-txi.md]: This is up for discussion. @fedekunze would you benefit from changing `Block` hash?

File: docs/references/architecture/adr-113-txi.md
Comment (melekes) [reply docs/references/architecture/adr-113-txi.md]: Pending discussion.

File: docs/references/architecture/adr-113-modular-transaction-hashing.md
Comment (sergio-mena) [inline docs/references/architecture/adr-113-modular-transaction-hashing.md:80]: How about `SetHashFn` ?

File: docs/references/architecture/adr-113-modular-transaction-hashing.md
Comment (sergio-mena) [inline docs/references/architecture/adr-113-modular-transaction-hashing.md:118]: That's a very important point. And I would add some reference to the existing light clients which will all probably get broken if the app dev's change the hashing function after genesis.

File: docs/references/architecture/adr-113-modular-transaction-hashing.md
Comment (sergio-mena) [inline docs/references/architecture/adr-113-modular-transaction-hashing.md:128]: If I am not missing something obvious, this would change the block format, forcing all chains upgrading to this version to hard-fork. Not an option at this stage.
We can't do this before we tackle Soft Upgrades (hopefully in Q3).

File: docs/references/architecture/adr-113-modular-transaction-hashing.md
Comment (sergio-mena) [inline docs/references/architecture/adr-113-modular-transaction-hashing.md:135]: Light clients of a chain MUST know the hash function used in that chain. Maybe, we should include the hashing algorithm in a filed in the genesis file.

File: docs/references/architecture/adr-113-modular-transaction-hashing.md
Comment (sergio-mena) [inline docs/references/architecture/adr-113-modular-transaction-hashing.md:147]: > since the proof contains `Hash`

Not sure I understand this part. Could you clarify?

File: docs/references/architecture/adr-113-modular-transaction-hashing.md
Comment (melekes) [reply docs/references/architecture/adr-113-modular-transaction-hashing.md:76]: Yes, I did. I found that having two separate things (object and a function is a cleaner approach)

Comment (sergio-mena) [reply docs/references/architecture/adr-113-modular-transaction-hashing.md:76]: Fair enough

File: docs/references/architecture/adr-113-modular-transaction-hashing.md
Comment (melekes) [reply docs/references/architecture/adr-113-modular-transaction-hashing.md:138]: You're right. We don't need to put it in the header.

File: docs/references/architecture/adr-113-modular-transaction-hashing.md
Comment (sergio-mena) [inline docs/references/architecture/adr-113-modular-transaction-hashing.md:155]: What does "5" mean here?

**Review by sergio-mena — APPROVED** [review]
🚀

File: docs/references/architecture/adr-113-modular-transaction-hashing.md
Comment (melekes) [reply docs/references/architecture/adr-113-modular-transaction-hashing.md:31]: It's a reference-style link (https://www.markdownguide.org/basic-syntax/#reference-style-links). The actual path is defined at the end of the document.

File: docs/references/architecture/adr-113-modular-transaction-hashing.md
Comment (melekes) [reply docs/references/architecture/adr-113-modular-transaction-hashing.md:76]: I agree, but the problem is that the standard `hash.Hash` https://pkg.go.dev/hash#Hash doesn't have a `String` method, so this will require our own interface, meaning we won't be able to use `crypto.Hash` from the Go stdlib. Again, nothing is wrong, but ideally, we'd use standard interfaces.

File: docs/references/architecture/adr-113-modular-transaction-hashing.md
Comment (melekes) [reply docs/references/architecture/adr-113-modular-transaction-hashing.md:157]: Go's stdlib `Hash` https://pkg.go.dev/crypto#Hash. We can also use strings like `SHA256` if that's more practical

File: docs/references/architecture/adr-113-modular-transaction-hashing.md
Comment (melekes) [reply docs/references/architecture/adr-113-modular-transaction-hashing.md]: I was looking at the wrong thing, sorry. Fixed

**Comment by melekes** [comment]
@fedekunze If we only allowed changing the hash function for transactions (how they are hashed in mempool and displayed, plus the Merkle root hash of transactions—`DataHash` in the block's header), would it suffice for your use case?

File: docs/references/architecture/adr-113-modular-transaction-hashing.md
Comment (cason) [inline docs/references/architecture/adr-113-modular-transaction-hashing.md:15]: Although we directly import `crypto/sha256` in many places:

```
$ grep crypto/sha256 . -R 2> /dev/null
./statesync/snapshots.go:	"crypto/sha256"
./crypto/merkle/tree.go:	"crypto/sha256"
./crypto/merkle/bench_test.go:	"crypto/sha256"
./crypto/bls12381/key_bls12381.go:	"crypto/sha256"
./crypto/secp256k1/secp256k1.go:	"crypto/sha256"
./crypto/tmhash/hash_test.go:	"crypto/sha256"
./crypto/tmhash/hash.go:	"crypto/sha256"
./crypto/tmhash/bench_test.go:	"crypto/sha256"
./crypto/hash.go:	"crypto/sha256"
./types/tx.go:	"crypto/sha256"
./test/e2e/app/state.go:	"crypto/sha256"
./p2p/conn/secret_connection.go:	"crypto/sha256"
./mempool/cache_test.go:	"crypto/sha256"
./mempool/mempool.go:	"crypto/sha256"
```

File: docs/references/architecture/adr-113-modular-transaction-hashing.md
Comment (cason) [inline docs/references/architecture/adr-113-modular-transaction-hashing.md:27]: It is also used in the mempool. Is it worth mentioning this?

File: docs/references/architecture/adr-113-modular-transaction-hashing.md
Comment (cason) [inline docs/references/architecture/adr-113-modular-transaction-hashing.md:49]: `HashFn` is a transaction hasher implementation?

File: docs/references/architecture/adr-113-modular-transaction-hashing.md
Comment (cason) [inline docs/references/architecture/adr-113-modular-transaction-hashing.md:58]: One solution I see in several places in the code base is:

1. Introduce an interface
2. Produce a standard implementation for the interface
3. Enable changing the implementation for the interface, which is essentially only used for testing

So I wonder why this wouldn't be a better solution.

File: docs/references/architecture/adr-113-modular-transaction-hashing.md
Comment (cason) [inline docs/references/architecture/adr-113-modular-transaction-hashing.md:80]: This can only be called before anything is started, right? Otherwise we have a real synchronization problem here.

File: docs/references/architecture/adr-113-modular-transaction-hashing.md
Comment (cason) [inline docs/references/architecture/adr-113-modular-transaction-hashing.md:107]: Again, this is for transactions or anything?

File: docs/references/architecture/README.md
Comment (cason) [inline docs/references/architecture/README.md:46]: The name here does not match the title of the next file.

**Review by cason — COMMENTED** [review]
A further review round.

Should we consider a prototype implementation to test the idea?

File: docs/references/architecture/adr-113-modular-transaction-hashing.md
Comment (melekes) [reply docs/references/architecture/adr-113-modular-transaction-hashing.md:112]: Clarified.

File: docs/references/architecture/adr-113-modular-transaction-hashing.md
Comment (melekes) [reply docs/references/architecture/adr-113-modular-transaction-hashing.md]: Correct.

File: docs/references/architecture/adr-113-modular-transaction-hashing.md
Comment (melekes) [reply docs/references/architecture/adr-113-modular-transaction-hashing.md]: It's an option that we pass to `NewNode` and forward to mempool, tx indexer, and tx RPC endpoints. However, the use case is rare, so I chose to opt for global variables instead.

File: docs/references/architecture/adr-113-modular-transaction-hashing.md
Comment (melekes) [reply docs/references/architecture/adr-113-modular-transaction-hashing.md:59]: https://github.com/cometbft/cometbft/pull/2241#discussion_r1633278671

My main argument against passing a hasher everywhere is that it's rare when somebody wants a different hashing function, so changing the interface of mempool, tx indexer and RPC endpoints is unjustified. But maybe I'm wrong. Please let me know what you think.

**Comment by cason** [comment]
I think we should also add a short paragraph mentioning that it is probably not possible/feasible to implement this via configuration parameter or command-line option, am I right? Because this could be seen as a trivial solution, but... unfeasible.

**Comment by cason** [comment]
Did we mention https://github.com/tendermint/tendermint/issues/6539 and https://github.com/tendermint/tendermint/pull/6773 here? It would be nice, also to see the discussions back then.

**Comment by melekes** [comment]
> I think we should also add a short paragraph mentioning that it is probably not possible/feasible to implement this via configuration parameter or command-line option, am I right? 

It's possible but I'm against this because

a) we're not sure this will be a final design/solution
b) we're optimizing for the default hashing function (by analogy, we may support different consensus algorithms, but we're optimizing for the default one)

**Pull request merged**
